### ISSUE TITLE
errors on debian stretch

### ISSUE THREAD

**Issue by tommyp-vr4** [v00d00/mediatomb#49]
Hi!  First time poster so please be gentle.  I'm trying to get mediatomb_v00d00 working on a headless debian-stretch. 
Part 1 - I tried cross compiling the ubuntu PPA (dated 3/14) into debian but it had a problem.  I tried compiling with the method on the main project page here and the error seems the be the same.  See the attached cmakeerror.log file.    
[CMakeError.log.txt](https://github.com/v00d00/mediatomb/files/849132/CMakeError.log.txt)

Part 2 - Late last year, Debian changed their mysqlclient-dev to default-mysqlclient-dev.  There were some BTS reported on the debian version of mediatomb saying that the debian/control file should be updated.  Can you please confirm which is the proper one I should be using?  

Part 3 - Not sure if this is an repacked ubuntu compiling artifact or not.  A few weeks back I successfully compiled the ubuntu ppa into debian.  I was able to get the mediatomb-daemon to work - but it only worked after I installed exim4.  The daemon stil worked after I removed exim4.  Any clue on what that is?  I compiled the debian MT and it worked without the exim4 installed.  

Part 4 - Is it worthwhile to add net-tools as a dependency?  The daemon needs it - and a CLI machine doesn't include net-tools by default.   

Sorry so many questions.  
Tommyp

**Comment by v00d00** [comment]
I dont really have any idea on Debian packaging, perhaps @stephen-czetty can help with this?

**Comment by tommyp-vr4** [comment]
Actually the log file from #1 shows the errors when I ran the commands from instructions on the main page (quick start build instructions).  The error I see closely tracks the repacked ubuntu package error too.  Is there some other or new dependency not listed?

**Comment by v00d00** [comment]
Can you post your cmake log from item 1?

**Comment by v00d00** [comment]
Also which version of CMake are you using?

**Comment by tommyp-vr4** [comment]
Did some more testing just a few minutes ago.  Seems that part of the problem is the ubuntu libupnp8 (yes I repacked it to a debian deb).  I installed the libupnp8 per the v00d00 script folder.  That went without a hitch.  I tried again to compile MT and it got further along.  The last of the console window said this.  I'm getting further though.

   dh_makeshlibs -O--buildsystem=cmake
   dh_shlibdeps -O--buildsystem=cmake
dpkg-shlibdeps: error: no dependency information found for /usr/lib/libupnp-1.8.so.10 (used by debian/mediatomb-common/usr/bin/mediatomb)
Hint: check if the library actually comes from a package.
dh_shlibdeps: dpkg-shlibdeps -Tdebian/mediatomb-common.substvars debian/mediatomb-common/usr/bin/mediatomb returned exit code 2
debian/rules:39: recipe for target 'binary' failed
make: *** [binary] Error 2
dpkg-buildpackage: error: fakeroot debian/rules binary gave error exit status 2
debuild: fatal error at line 1116:
dpkg-buildpackage -rfakeroot -us -uc failed

**Comment by jcowgill** [comment]
>dpkg-shlibdeps: error: no dependency information found for /usr/lib/libupnp-1.8.so.10

This is a bug in your libupnp package. This dependency information has to be provided by the library package but it apparently isn't there.

I have been packaging 1.8 here if you want to try it (on the `upnp-1.8` branch). I think it's almost done except for a few copyright related bits:
https://anonscm.debian.org/cgit/users/jcowgill/pupnp-1.8.git

**Comment by tommyp-vr4** [comment]
@jcowgill.  Can you send me an offline message on how to compile it?  I'm a learning newbee (I ran a git clone, cd'd into the directory and when I ran debuild -us -uc I got an error). Also - don't forget to update your debian/control file.  Looks like it references v1.6.  
@v00d00.  See my very first post for the error log file (I ranamed it with a txt suffix).
You guys rock.  This is my first exposure to providing feedback to the open source community.

**Comment by jcowgill** [comment]
@tommyp-vr4 Try:
```
git checkout upnp-1.8
dpkg-buildpackage -b -us -uc
```

**Comment by tommyp-vr4** [comment]
I feel horrible for contaminating v00d00's mediatomb issues message thread area.  Seems most of my problems are due to libupnpv1.8.  Keep in mind if I install the way posted in the mediatomb/scripts/install-pupnp18.sh, I still can't compile mediatomb.  
@jcowgill Didn't have luck rebuilding upnp-1.8.  Since I'm a stupid noob, I restarted from scratch by rebuilding my VM just to be sure I didn't do anything stupid.  Not sure if it matters, but I'm using debian stretch, cmake is version 3.7.2-1.   See below for the errors (can't locate any log files to attach - hope the console text is provides enough guidance).  
my commands:
* git clone https://anonscm.debian.org/cgit/users/jcowgill/pupnp-1.8.git
* CD into libupnp-1.8
* dpkg-buildpackage -b -us -uc

Here's the results:

dpkg-buildpackage: info: host architecture amd64
 dpkg-source --before-build pupnp-1.8
dpkg-source: info: applying 01-debian-md5-licence.patch
dpkg-source: info: applying 07-neaten-debug.patch
patching file upnp/src/api/upnpdebug.c
Hunk #2 FAILED at 61.
Hunk #3 succeeded at 90 (offset 2 lines).
Hunk #4 succeeded at 151 (offset 2 lines).
Hunk #5 succeeded at 181 (offset 2 lines).
Hunk #6 succeeded at 191 (offset 2 lines).
1 out of 6 hunks FAILED
patching file upnp/src/gena/gena_ctrlpt.c
Hunk #1 FAILED at 79.
Hunk #2 succeeded at 107 (offset -1 lines).
Hunk #3 succeeded at 525 (offset -6 lines).
Hunk #4 FAILED at 639.
2 out of 4 hunks FAILED
patching file upnp/src/inc/gena.h
patching file upnp/src/inc/upnpapi.h
patching file upnp/src/gena/gena_device.c
Hunk #17 succeeded at 1263 (offset 5 lines).
Hunk #18 succeeded at 1437 (offset 4 lines).
patching file upnp/src/genlib/net/http/httpreadwrite.c
Hunk #1 succeeded at 371 (offset 60 lines).
Hunk #2 succeeded at 403 (offset 60 lines).
Hunk #3 succeeded at 775 (offset 60 lines).
Hunk #4 succeeded at 1501 (offset -158 lines).
Hunk #5 succeeded at 1523 (offset -158 lines).
patching file upnp/src/genlib/net/http/webserver.c
Hunk #1 succeeded at 1053 (offset 4 lines).
Hunk #2 succeeded at 1420 (offset 6 lines).
patching file upnp/src/ssdp/ssdp_server.c
patching file upnp/src/genlib/net/http/httpparser.c
Hunk #1 succeeded at 2195 (offset 4 lines).
patching file upnp/src/inc/httpparser.h
Hunk #1 succeeded at 480 (offset 4 lines).
dpkg-source: info: the patch has fuzz which is not allowed, or is malformed
dpkg-source: info: if patch '07-neaten-debug.patch' is correctly applied by quilt, use 'quilt refresh' to update it
dpkg-source: info: restoring quilt backup files for 07-neaten-debug.patch
dpkg-source: error: LC_ALL=C patch -t -F 0 -N -p1 -u -V never -E -b -B .pc/07-neaten-debug.patch/ --reject-file=- < pupnp-1.8/debian/patches/07-neaten-debug.patch gave error exit status 1
dpkg-buildpackage: error: dpkg-source --before-build pupnp-1.8 gave error exit status 2

**Comment by jcowgill** [comment]
@tommyp-vr4 You forgot to `git checkout upnp-1.8`, though I guess I could just push everything to master.

**Comment by tommyp-vr4** [comment]
The git stuff is new to me.  Ran the git checkout and the debs were made and installed them.  Thanks for the tip.  Following that, I ran a few wget commands to make a local copy of steve czetty's ppa of the mediatomb files and ran a "dpkg-source -x mediatomb_0.13.0*.dsc" command to create the directories.  Ended up that I needed to change the debian/control file to specify "libupnp-1.8-dev" instead of "libupnp8-dev".  Guess it's a debian and/or ubuntu naming standard that isn't quite standardized yet???  Not a big deal, I am learning what to change to get things to work.  I cd'd into the working directory and ran a "debuild -us -uc" command.  See attached error log.  At quick glance, the errors look the same.  See attached build log.  
I seem to recall v00d00 posting something on the libupnp8 github site about incorporating ipv6 into mediatomb.  Are the right commandline switches in place to enable ipv6? Or does it not matter?  Just trying to throw out some ideas.  I was able to build mediatomb using steve czetty's 20170306 git date.  Something happened between 0306 to 0314.  Not sure what.  Perhaps that is a clue?  
[mediatomb_0.13.0+alpha~git20170314-1_amd64.build.txt](https://github.com/v00d00/mediatomb/files/851646/mediatomb_0.13.0.alpha.git20170314-1_amd64.build.txt)

**Comment by tommyp-vr4** [comment]
And here are some log files when I compile using git clone of mediatomb / cmake (my previous message had log files using a different method).  Hope these files help out.  :)
[CMakeError.log.txt](https://github.com/v00d00/mediatomb/files/851684/CMakeError.log.txt)
[CMakeOutput.log.txt](https://github.com/v00d00/mediatomb/files/851685/CMakeOutput.log.txt)

**Comment by jcowgill** [comment]
```
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
UPNP_THREADUTIL_LIBRARY (ADVANCED)
```

This is a mediatomb bug triggered by mrjimenez/pupnp#22 and mrjimenez/pupnp#25. In upnp git, the threadutil library no longer exists and to make the packaging easier I applied that change to my Debian packaging as well.

If you try the fix from #50 it should work.

**Comment by tommyp-vr4** [comment]
Thanks @jcowgill !!!   We're getting closer.  Made manual corrections to that make file and ran the long make command.  Was greeted with a a cannot find libav/ffmpeg or something or other error.  I'm not sure what to download for debian to clear that error.  I can provide whatever file you need to help fix that.  
So I ended up modifying the czetty's ppa code, it compiled (yay) and it installed (another yay).  Set the proper settings in /etc/default/mediatomb and ran the systemctl enable mediatomb so it autolaunches.  Seemed to do the trick.... the daemon works, but it seems to only work in a desktop environment.  :(

So now, I'm back to where I left off from a few days ago.  My plan is to run MT on a headless debian stretch machine (i.e. no DE).  So, testing resumes on a headless VM.  Kind of tricky doing all by ssh, but doable I guess.  I copied the debs (jcowgill's libupnp-1.8 and czetty's debian compiled mediatomb) into a VM folderstructure that doesn't have a DE.  The other requisites are already available in debian stretch. All installed OK, apt is not complaining.  However, I am unable to get the daemon to autolaunch.  I'm not sure what files I can provide to help out.  The /var/log/mediatomb file has this error:
2017-03-17 18:06:01    INFO: Configuration check succeeded.
Exception raised in [/home/tom/mediatomb_1/mediatomb-0.13.0+alpha~git20170314/s$
2017-03-17 18:06:01   ERROR: main: upnp error -203
2017-03-17 18:06:01   ERROR: Could not bind to socket.
2017-03-17 18:06:01    INFO: Please check if another instance of MediaTomb or
2017-03-17 18:06:01    INFO: another application is running on port 50000.
2017-03-17 18:06:01   ERROR: upnp_cleanup: UpnpUnRegisterRootDevice failed
Exception raised in [/home/tom/mediatomb_1/mediatomb-0.13.0+alpha~git20170314/s$

However, if I run a /etc/init.d/mediatomb start as root, then mediatomb starts.  Very odd.  If I compile the debian/sid mediatomb the daemon works just fine.  Any further tips?  I wish I was smart enough to make the debian installer (i.e. dsc and whatnot) myself instead of relying on cross compiling from ubuntu.

**Comment by tommyp-vr4** [comment]
I ran some experiments this morning.  Two different installation methods/packages.  I'll try to be specific because I know how troubleshooting is.  If I am too wordy, let me know and I'll be shorter with my words.  
Package 1 - v00d00 mediatomb with @jcowgill 's libupnp8.  These packages work if there is a DE installed (and exim4 for some reason) but does NOT run on a CLI/headless machine.  Yes, that is stated correctly - a virgin DE VM wtih the packages installed also need exim4 to be installed for the daemon to work.  Not sure why but I think this provides a clue to the problems.   This package does not work in a cli/headless VM.  
Package 2 - Debian sid (the unstable branch) that is compiled into debian stretch. This runs fine on a DE machine AND a CLI/headless machine.  

Test sequence on a cli/headless VM.  
Result 1.  Install the packages defined in package 1 above.  Set configuration and set autostart (systemctl enable mediatomb).  Set the mediatomb.service with -D option for debugging.  Reboot & NG.  
Result 2.  Install packages defined in package 2 above.  Set configuration and set autostart and debug switch.  Reboot and it works great.  

There is some sort of snafu that prevents the v00d00 mediatomb to run as a service/daemon.  I hope the two mediatomb log files can help identify the problem.  One file is the working "sid" configuration and the other is the broken "v00d00" configuration.  
[mediatomb.sid.txt](https://github.com/v00d00/mediatomb/files/852448/mediatomb.sid.txt)
[mediatomb.v00d00.txt](https://github.com/v00d00/mediatomb/files/852447/mediatomb.v00d00.txt)

**Comment by tommyp-vr4** [comment]
I dug a bit deeper and tinkered to find a solution to my mediatomb problems on debian stretch. Remember I'm a novice and this info could be bad - but it seems to work fine for me.  These fixes work on a CLI or headless debian stretch machine.  A graphical DE is more forgiving with pre-reqs and settings, but the CLI needed some love.  

1.  To fix the mediatomb daemon that does not autolaunch I had to modify the /lib/systemd/system/mediatomb.service as follows:
Part a - In the [Unit] section - change the After line to read After=network.target  
The NetworkManager-wait-online.service seems to be partial cause of problems.  Samba, for example, uses After=network.target so this confirms things a bit.  
Part b - At the end of the [Service] section - add a line that says Restart=on-failure
Bottom line is that the systemd doesn't know when the network is running, and a quick google search said the same thing. Since systemd doesn't know exactly when the network is up and running, then MT will attempt to relaunch itself if it doesn't start.  

2.  The "net-tools" is a prerequisite that should be added to your list of tools.  It is NOT included in a headless debian net install and am not sure of other CLI only distros.  The daemon depends on it - and it's not called out. I had discovered after manually parsing through the mediatomb.service file, and found some debian and other distros bug reports saying the same thing a few years back. Not sure why software maintainers never addressed it.   

3.  default-mysqlclient-dev appears to be the correct version for debian stretch.  The mysqlclient-dev was renamed to default-mysqlclient-dev late last year.  This too was a debian bug report that was never addressed.  

4.  On a big sidenote - I highly recommend a standard libupnp8 or libupnp-1.8 package naming convention.  Two different people and two different package names.  We should nip this in the bud as this library file is relatively new. I know it's not a mediatomb issue, but maybe you guys have some say.  

My issues are now closed (and I do not know how to mark a "solved" button on this thread.  I left this message here to help others out (and hopefully Steve Czetty reads this and rolls these suggested updates into his PPA).  Or perhaps someone will create an official debian folder for mediatomb_v00d00 so any debian user can make a deb.
